torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,455 @@
+class BatchDataBuffer {
+ public:
+  using UnwrappedBatchType = UnwrappedBatch;
+  using BatchType = torch::optional<UnwrappedBatchType>;
+  using BatchRequestType = typename ExampleSampler::BatchRequestType;
+
+  BatchDataBuffer(
+      size_t num_chunks,
+      size_t batch_size,
+      ExampleSampler example_sampler,
+      bool ignore_empty_chunk,
+      size_t queue_capacity)
+      : remaining_chunk_count_(num_chunks),
+        batch_size_(batch_size),
+        example_sampler_(std::move(example_sampler)),
+        ignore_empty_chunk_(ignore_empty_chunk),
+        queue_capacity_(queue_capacity) {}
+
+  /// Return batch data from the queue. Called from the ChunkDataset main
+  /// thread.
+  BatchType get_batch() {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    cv_read_.wait(lock, [this] {
+      // wait till there is available data in the queue or if all chunks are
+      // loaded (i.e. the dataset is exhausted for this epoch)
+      return (
+          this->total_example_count_in_queue_ >= batch_size_ ||
+          this->remaining_chunk_count_ == 0);
+    });
+    if (batch_queue_.empty()) {
+      lock.unlock();
+      AT_ASSERT(remaining_chunk_count_ == 0);
+
+      // All batches have been retrieved. Return an empty batch.
+      return nullopt;
+    }
+
+    auto batch = batch_queue_.front();
+    batch_queue_.pop();
+    if (batch.exception) {
+      throw WorkerException(batch.exception);
+    }
+
+    total_example_count_in_queue_ -= batch.batch_data.size();
+    lock.unlock();
+    cv_write_.notify_all();
+
+    return batch.batch_data;
+  }
+
+  // skip one chunk
+  void skip_chunk() {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    AT_ASSERT(remaining_chunk_count_ > 0);
+    remaining_chunk_count_--;
+    lock.unlock();
+    cv_read_.notify_all();
+  }
+
+  /// Push preloaded chunks to batch queue. Called from the ChunkDataset worker
+  /// threads.
+  void add_chunk_data(UnwrappedBatchType data) {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    cv_write_.wait(lock, [this] {
+      // stop loading if we have preloaded enough data.
+      return this->total_example_count_in_queue_ < this->queue_capacity_ || stop_;
+    });
+
+    if (stop_){
+      // When stop_ is true, it means this current thread needs to be tore down.
+      // Return without any further processing.
+      lock.unlock();
+      return;
+    }
+
+    auto data_size = data.size();
+    auto remaining_size = data_size;
+    example_sampler_.reset(data_size);
+
+    auto fill_batch = [&](size_t example_count, UnwrappedBatchType& batch) {
+      auto batch_example_indices = this->example_sampler_.next(example_count);
+      AT_ASSERT(
+          batch_example_indices &&
+          batch_example_indices.value().size() == example_count)
+      BatchRequestType indices = batch_example_indices.value();
+      for (size_t i : indices) {
+        batch.emplace_back(std::move(data[i]));
```

Comment:
I know the Sampler interface states that it should return indices in this range, but in case someone has a bug in their implementation this will cause invalid memory accesses. It doesn't cost much to assert that the indices are within range, and greatly increases safety.